### PR TITLE
Paulcallen/shutdown

### PR DESCRIPTION
**Pull request by paulcallen**

### CODE REVIEW

File: include/myst/thread.h [hunk 1]
```diff
@@ -419,7 +419,7 @@ void myst_fork_exec_futex_wake(myst_thread_t* thread);
 
 size_t myst_kill_thread_group();
 
-long myst_have_child_forked_processes(myst_thread_t* process);
+bool myst_have_child_forked_processes(myst_thread_t* process);
```
Comment (mikbras) [inline include/myst/thread.h:422]: Nice.

File: kernel/enter.c [hunk 2]
```diff
@@ -860,47 +862,78 @@ int myst_enter_kernel(myst_kernel_args_t* args)
...
         }
 
-        /* free all non-process threads, waiting for all other threads to
-         * shutdown at the same time. Our thread has not been marked as a zombie
-         * yet. */
+        /* Wait for all child threads to shutdown */
         {
-            myst_thread_t* t = thread->group_next;
-            while (t)
+            myst_assume(thread->group_prev == NULL);
```
Comment (mikbras) [inline kernel/enter.c:867]: Good. We'll see if this crashes. :)

File: kernel/enter.c [hunk 3]
```diff
@@ -860,47 +862,78 @@ int myst_enter_kernel(myst_kernel_args_t* args)
...
+        /* Remove ourself from /proc/<pid> so other processes know we have gone
+         * if they check */
+        procfs_pid_cleanup(thread->pid);
+
+        /* Send SIGHUP to all other active processes */
+        myst_send_sighup_child_processes(thread);
+
+        /* Put the thread on the zombie list, although no one will be doing a
+         * waitpid on this top-level process, but it will make it consistent */
+        //        myst_zombify_thread(thread);
```
Comment (mikbras) [inline kernel/enter.c:893]: Where does the thread get zombified now?

File: kernel/thread.c [hunk 4]
```diff
@@ -567,7 +567,7 @@ long myst_wait(
                         p->znext->zprev = p->zprev;
 
                     if (_zombies_tail == p)
-                        _zombies_tail = p->group_prev;
+                        _zombies_tail = p->zprev;
```
Comment (mikbras) [inline kernel/thread.c:570]: Good catch!

File: tests/shutdown/child.c [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+    else if (
+        strcmp("parent-wait-child-spawn-exit-with-sighup-handler", argv[1]) ==
+        0)
+        test2(argv[1]);
+    else
+        assert("invalid option" == NULL);
+
+    printf("goodbye from %s %s\n", argv[0], argv[1]);
+    return 0;
+}
```
Comment (mikbras) [inline tests/shutdown/child.c:47]: Looks like code-format could be run on this file.

**Review by mikbras — APPROVED** [review]
Looks great. Just one question about commented out "zombify-thread" call. Do we need it.

File: kernel/enter.c (hunk 3 above)
Comment (paulcallen) [reply kernel/enter.c:893]: this is the initial process thread and you removed that code. I'll remove this line

**Pull request merged**